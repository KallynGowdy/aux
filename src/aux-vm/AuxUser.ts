import { User } from '@casual-simulation/causal-trees';

export interface AuxUser extends User {
    isGuest: boolean;
    channelId: string;
}
