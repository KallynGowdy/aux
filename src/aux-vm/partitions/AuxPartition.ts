import {
    AuxCausalTree,
    BotsState,
    BotAction,
    merge,
    Bot,
    UpdatedBot,
} from '@casual-simulation/aux-common';
import {
    RealtimeCausalTree,
    DeviceAction,
    StatusUpdate,
    RemoteAction,
    User,
} from '@casual-simulation/causal-trees';
import { Observable, SubscriptionLike } from 'rxjs';
import { StoredAux } from '../StoredAux';

/**
 * Defines an interface that maps Bot IDs to their corresponding partitions.
 */
export interface AuxPartitions {
    '*': AuxPartition;
    [key: string]: AuxPartition;
}

/**
 * Defines a set of valid partition types.
 */
export type AuxPartition =
    | CausalTreePartition
    | MemoryPartition
    | RemoteCausalTreePartition
    | CausalRepoPartition
    | RemoteCausalRepoPartition
    | LocalStoragePartition;

/**
 * Base interface for partitions.
 *
 * Partitions are basically a backing store for Aux State.
 * They allow working on and manipulating bots that are stored in multiple different places.
 */
export interface AuxPartitionBase extends SubscriptionLike {
    /**
     * Whether the partition is private or not.
     * If true, then the partition will be skipped when exporting state.
     * If false, then the partition will be included when exporting state.
     */
    private: boolean;

    /**
     * Applies the given events to the partition.
     * Returns events that should be sent as local events.
     * @param events The events to apply.
     */
    applyEvents(events: BotAction[]): Promise<BotAction[]>;

    /**
     * Sends the given events to the targeted device.
     * @param events The events to send.
     */
    sendRemoteEvents?(events: RemoteAction[]): Promise<void>;

    /**
     * Sets the user that the partition should use.
     * @param user
     */
    setUser?(user: User): Promise<void>;

    /**
     * Sets the grant that the partition should use.
     * @param grant
     */
    setGrant?(grant: string): Promise<void>;

    /**
     * Tells the partition to connect to it's backing store.
     */
    connect(): void;

    /**
     * Gets an observable list that resolves whenever
     * a bot is added to this partition.
     */
    onBotsAdded: Observable<Bot[]>;

    /**
     * Gets an observable list that resolves whenever
     * a bot is removed from this partition.
     */
    onBotsRemoved: Observable<string[]>;

    /**
     * Gets an observable list that resolves whenever
     * a bot is updated in this partition.
     */
    onBotsUpdated: Observable<UpdatedBot[]>;

    /**
     * Gets an observable list of errors from the partition.
     */
    onError: Observable<any>;

    /**
     * Gets the observable list of remote events from the partition.
     */
    onEvents: Observable<DeviceAction[]>;

    /**
     * Gets the observable list of status updates from the partition.
     */
    onStatusUpdated: Observable<StatusUpdate>;
}

/**
 * Defines a causal repo partition.
 */
export interface CausalRepoPartition extends AuxPartitionBase {
    type: 'causal_repo';

    state: BotsState;
}

/**
 * Defines a remote causal repo partition.
 * That is, a partition that was loaded from a remote server.
 */
export interface RemoteCausalRepoPartition extends CausalRepoPartition {
    /**
     * Gets or sets whether the partition has been forced offline.
     */
    forcedOffline: boolean;
}

/**
 * Defines a causal tree partition.
 */
export interface CausalTreePartition extends AuxPartitionBase {
    type: 'causal_tree';

    /**
     * The causal tree for the partition.
     */
    tree: AuxCausalTree;
}

/**
 * Defines a remote causal tree partition.
 * That is, a causal tree partition that was loaded from a remote server.
 */
export interface RemoteCausalTreePartition extends CausalTreePartition {
    /**
     * The realtime causal tree that represents the partition connnection.
     */
    sync: RealtimeCausalTree<AuxCausalTree>;

    /**
     * Forks the current causal tree to a channel with the given ID.
     * @param newId The ID of the new channel.
     * @param events The events that should be applied to the newly forked causal tree. Use this to apply some sort of reset operation.
     */
    fork(newId: string, events: BotAction[]): Promise<void>;

    setUser(user: User): Promise<void>;
    setGrant(grant: string): Promise<void>;

    /**
     * Gets or sets whether the partition has been forced offline.
     */
    forcedOffline: boolean;
}

/**
 * Defines a memory partition.
 */
export interface MemoryPartition extends AuxPartitionBase {
    type: 'memory';

    /**
     * The current state for the partition.
     */
    state: BotsState;
}

/**
 * Defines a local storage partition.
 * Needs to run on the main thread.
 */
export interface LocalStoragePartition extends AuxPartitionBase {
    type: 'local_storage';

    /**
     * The namespace that bots should be stored under.
     */
    namespace: string;
}

/**
 * Gets the bots state from the given partition.
 * @param partition The partition.
 */
export function getPartitionState(partition: AuxPartition): BotsState {
    if (partition.type === 'causal_tree') {
        return partition.tree.value;
    } else {
        return partition.state;
    }
}

/**
 * Iterates the given partitions.
 * @param partitions The partitions to iterate.
 */
export function* iteratePartitions(partitions: AuxPartitions) {
    for (let key in partitions) {
        if (!partitions.hasOwnProperty(key)) {
            continue;
        }

        yield [key, partitions[key]] as const;
    }
}
