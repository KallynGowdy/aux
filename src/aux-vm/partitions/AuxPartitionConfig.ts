import { BotsState, AuxCausalTree } from '@casual-simulation/aux-common';
import { User } from '@casual-simulation/causal-trees';
import { CausalRepoClient } from '@casual-simulation/causal-trees/core2';
import {
    AuxPartition,
    ProxyBridgePartition,
    MemoryPartition,
} from './AuxPartition';

/**
 * Defines a set of options for configuring partitioning of bots.
 * Bot IDs are mapped to
 */
export interface AuxPartitionConfig {
    shared: PartitionConfig;
    [key: string]: PartitionConfig;
}

/**
 * Defines a partition config.
 * That is, a config which specifies how to build a partition.
 */
export type PartitionConfig =
    | RemoteCausalTreePartitionConfig
    | CausalTreePartitionConfig
    | CausalRepoPartitionConfig
    | RemoteCausalRepoPartitionConfig
    | CausalRepoHistoryClientPartitionConfig
    | CausalRepoClientPartitionConfig
    | MemoryPartitionStateConfig
    | MemoryPartitionInstanceConfig
    | ProxyPartitionConfig
    | ProxyClientPartitionConfig
    | LocalStoragePartitionConfig;

/**
 * Defines a base interface for partitions.
 */
export interface PartitionConfigBase {
    /**
     * Whether the partition is private.
     * If true, then the contents of the partition should not be exported.
     * If false, then the bot state in the partition is exportable.
     * Defaults to false.
     */
    private?: boolean;
}

/**
 * Defines a memory partition.
 * That is, a configuration that specifies that bots should be stored in memory.
 */
export interface MemoryPartitionConfig extends PartitionConfigBase {
    type: 'memory';
}

export interface MemoryPartitionStateConfig extends MemoryPartitionConfig {
    /**
     * The initial state for the memory partition.
     */
    initialState: BotsState;
}

export interface MemoryPartitionInstanceConfig extends MemoryPartitionConfig {
    partition?: MemoryPartition;
}

/**
 * Defines a partition that proxies requests from the engine to the given partition instance.
 * Basically gives a way to run a partition on the main thread instead of in a background thread.
 * Useful for storing data using APIs that are only available to the main thread.
 */
export interface ProxyPartitionConfig extends PartitionConfigBase {
    type: 'proxy';

    /**
     * The partition that should be used.
     */
    partition: AuxPartition;
}

/**
 * Defines a partition that is able to proxy requests from the engine to the given partition bridge.
 */
export interface ProxyClientPartitionConfig extends PartitionConfigBase {
    type: 'proxy_client';

    /**
     * The port that should be used for messages.
     */
    port: MessagePort;
}

/**
 * Defines a partition that stores data in local storage.
 */
export interface LocalStoragePartitionConfig extends PartitionConfigBase {
    type: 'local_storage';

    /**
     * The namespace that the partition should store bots under.
     */
    namespace: string;
}

/**
 * Defines a causal tree partition.
 * That is, a configuration that specifies that bots should be stored in a causal tree.
 */
export interface CausalTreePartitionConfig extends PartitionConfigBase {
    type: 'causal_tree';

    /**
     * The tree to use.
     */
    tree: AuxCausalTree;

    /**
     * The ID of the tree.
     */
    id: string;
}

/**
 * Defines a causal tree partition that uses the new Causal Repo API.
 */
export interface CausalRepoPartitionConfig extends PartitionConfigBase {
    type: 'causal_repo';
}

/**
 * Defines a causal tree partition that uses the new Causal Repo API.
 */
export interface CausalRepoClientPartitionConfig extends PartitionConfigBase {
    type: 'causal_repo_client';

    /**
     * The branch to load.
     */
    branch: string;

    /**
     * The client that should be used to connect.
     */
    client: CausalRepoClient;

    /**
     * Whether the partition should be loaded in read-only mode.
     */
    readOnly?: boolean;
}

/**
 * Defines a causal tree partition that uses the new Causal Repo API.
 */
export interface RemoteCausalRepoPartitionConfig extends PartitionConfigBase {
    type: 'remote_causal_repo';

    /**
     * The branch to load.
     */
    branch: string;

    /**
     * The host that the branch should be loaded from.
     */
    host: string;

    /**
     * Whether the partition should be loaded in read-only mode.
     */
    readOnly?: boolean;
}

/**
 * Defines a causal repo partition that loads history for a branch.
 */
export interface CausalRepoHistoryClientPartitionConfig
    extends PartitionConfigBase {
    type: 'causal_repo_history_client';

    /**
     * The branch to load history from.
     */
    branch: string;

    /**
     * The client that should be used to load the history.
     */
    client: CausalRepoClient;
}

/**
 * Defines a remote causal tree partition.
 * That is, a configuration that specifies that bots should be stored in a causal tree
 * which is loaded from a remote server.
 */
export interface RemoteCausalTreePartitionConfig extends PartitionConfigBase {
    type: 'remote_causal_tree';

    /**
     * The ID of the tree.
     */
    id: string;

    /**
     * The host that should be connected to.
     */
    host: string;

    /**
     * The name of the tree to load.
     */
    treeName: string;
}
