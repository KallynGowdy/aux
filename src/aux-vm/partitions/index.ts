export * from './AuxPartition';
export * from './AuxPartitionConfig';
export * from './MemoryPartition';
export * from './AuxPartitionFactories';
export * from './CausalTreePartition';
export * from './LocalCausalTreePartition';
export * from './CausalTree2Partition';
