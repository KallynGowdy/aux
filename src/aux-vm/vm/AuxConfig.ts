import {
    FormulaLibraryOptions,
    AuxPartitionConfig,
} from '@casual-simulation/aux-common';

/**
 * Defines the possible configuration options for a simulation.
 */
export interface AuxConfig {
    config: AuxConfigParameters;

    /**
     * Defines the partitioning structure for bots.
     */
    partitions: AuxPartitionConfig;
}

export interface AuxConfigParameters {
    version: string;
    versionHash: string;
    builder?: string;
    device?: FormulaLibraryOptions['device'];
}

/**
 * Creates a set of formula library options for the given aux config.
 * @param config The config.
 */
export function buildFormulaLibraryOptions(
    config: AuxConfigParameters
): FormulaLibraryOptions {
    if (!config) {
        return null;
    }
    let options: FormulaLibraryOptions = {
        config: {},
        version: buildVersionNumber(config),
    };

    if (config.device) {
        options.device = config.device;
    }

    return options;
}

export function buildVersionNumber(config: AuxConfigParameters) {
    if (!config) {
        return null;
    }
    return {
        hash: config.versionHash,
        ...parseVersionNumber(config.version),
    };
}

/**
 * Parses the given version number.
 * @param version The version number.
 */
export function parseVersionNumber(version: string) {
    if (!version) {
        return {
            version: null,
            major: null,
            minor: null,
            patch: null,
        };
    }
    const versionRegex = /^v(\d+)\.(\d+)\.(\d+)$/i;
    const [str, major, minor, patch] = versionRegex.exec(version);

    return {
        version: str,
        major: parseInt(major),
        minor: parseInt(minor),
        patch: parseInt(patch),
    };
}
