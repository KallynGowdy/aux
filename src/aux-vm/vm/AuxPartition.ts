import {
    AuxCausalTree,
    BotsState,
    BotAction,
    merge,
} from '@casual-simulation/aux-common';

/**
 * Defines a set of options for configuring partitioning of bots.
 * Bot IDs are mapped to
 */
export interface AuxPartitionConfig {
    '*': PartitionConfig;
    [key: string]: PartitionConfig;
}

/**
 * Defines a partition config.
 * That is, a config which specifies how to build a partition.
 */
export type PartitionConfig = CausalTreePartitionConfig | MemoryPartitionConfig;

/**
 * Defines a memory partition.
 * That is, a configuration that specifies that bots should be stored in memory.
 */
export interface MemoryPartitionConfig {
    type: 'memory';

    /**
     * The initial state for the memory partition.
     */
    initialState: BotsState;
}

/**
 * Defines a causal tree partition.
 * That is, a configuration that specifies that bots should be stored in a causal tree.
 */
export interface CausalTreePartitionConfig {
    type: 'causal_tree';

    /**
     * The ID of the tree.
     */
    id: string;

    /**
     * The host that should be connected to.
     */
    host: string;

    /**
     * The name of the tree to load.
     */
    treeName: string;
}

/**
 * Defines an interface that maps Bot IDs to their corresponding partitions.
 */
export interface AuxPartitions {
    '*': Partition;
    [key: string]: Partition;
}

export type Partition = CausalTreePartition | MemoryPartition;

/**
 * Defines a causal tree partition.
 */
export interface CausalTreePartition {
    type: 'causal_tree';

    /**
     * The causal tree for the partition.
     */
    tree: AuxCausalTree;
}

/**
 * Defines a memory partition.
 */
export interface MemoryPartition {
    type: 'memory';

    /**
     * The current state for the partition.
     */
    state: BotsState;
}

/**
 * Gets the bots state from the given partition.
 * @param partition The partition.
 */
export function getPartitionState(partition: Partition): BotsState {
    if (partition.type === 'causal_tree') {
        return partition.tree.value;
    } else {
        return partition.state;
    }
}

export async function applyEvents(
    partition: Partition,
    events: BotAction[]
): Promise<void> {
    if (partition.type === 'causal_tree') {
        await partition.tree.addEvents(events);
    } else {
        for (let event of events) {
            applyEventToMemoryPartition(partition, event);
        }
    }
}

export function applyEventToMemoryPartition(
    partition: MemoryPartition,
    event: BotAction
) {
    if (event.type === 'add_bot') {
        partition.state = Object.assign({}, partition.state, {
            [event.bot.id]: event.bot,
        });
    } else if (event.type === 'remove_bot') {
        let { [event.id]: removedBot, ...state } = partition.state;
        partition.state = state;
    } else if (event.type === 'update_bot') {
        partition.state = merge(partition.state, {
            [event.id]: event.update,
        });
    }
}
