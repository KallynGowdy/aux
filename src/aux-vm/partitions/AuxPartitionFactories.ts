import {
    RemoteCausalTreePartitionConfig,
    AuxPartitionConfig,
    PartitionConfig,
    MemoryPartitionConfig,
} from './AuxPartitionConfig';
import { AuxPartition, MemoryPartition } from './AuxPartition';
import {
    BotsState,
    BotAction,
    Bot,
    UpdatedBot,
} from '@casual-simulation/aux-common';
import { Observable, Subject } from 'rxjs';

export type AuxPartitionFactory = (
    config: PartitionConfig
) => Promise<AuxPartition> | AuxPartition;

/**
 * Creates an AUX Partition from the given list of factory functions.
 * The first factory function that returns a partition is the partition
 * that gets returned.
 * @param config The config which indicates the type of partition to create.
 * @param factories The factory functions.
 */
export async function createAuxPartition(
    config: PartitionConfig,
    ...factories: AuxPartitionFactory[]
): Promise<AuxPartition> {
    for (let factory of factories) {
        let result = await factory(config);
        if (result) {
            return result;
        }
    }

    return undefined;
}
