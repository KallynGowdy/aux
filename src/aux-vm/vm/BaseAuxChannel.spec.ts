import { BaseAuxChannel, filterAtom } from './BaseAuxChannel';
import {
    RealtimeCausalTree,
    LocalRealtimeCausalTree,
    storedTree,
    site,
    AuthorizationMessage,
    USERNAME_CLAIM,
    DEVICE_ID_CLAIM,
    SESSION_ID_CLAIM,
    RemoteAction,
    DeviceAction,
    remote,
    DeviceInfo,
    ADMIN_ROLE,
    SERVER_ROLE,
    RealtimeCausalTreeOptions,
    atom,
    atomId,
} from '@casual-simulation/causal-trees';
import {
    AuxCausalTree,
    GLOBALS_BOT_ID,
    createBot,
    botAdded,
    botRemoved,
    bot,
    del,
    tag,
    value,
    DEFAULT_USER_DELETION_TIME,
    browseHistory,
} from '@casual-simulation/aux-common';
import { AuxUser } from '../AuxUser';
import { AuxConfig } from './AuxConfig';
import { AuxPartition } from '../partitions/AuxPartition';
import {
    PartitionConfig,
    MemoryPartitionConfig,
} from '../partitions/AuxPartitionConfig';
import { createAuxPartition, createLocalCausalTreePartitionFactory } from '..';
import uuid from 'uuid/v4';
import { createMemoryPartition } from '../partitions';
import merge from 'lodash/merge';

const uuidMock: jest.Mock = <any>uuid;
jest.mock('uuid/v4');

const nowMock = (Date.now = jest.fn());

console.log = jest.fn();
console.warn = jest.fn();
console.error = jest.fn();

describe('BaseAuxChannel', () => {
    let channel: AuxChannelImpl;
    let user: AuxUser;
    let device: DeviceInfo;
    let config: AuxConfig;
    let tree: AuxCausalTree;

    beforeEach(async () => {
        user = {
            id: 'userId',
            username: 'username',
            isGuest: false,
            name: 'name',
            token: 'token',
        };
        device = {
            claims: {
                [USERNAME_CLAIM]: 'username',
                [DEVICE_ID_CLAIM]: 'deviceId',
                [SESSION_ID_CLAIM]: 'sessionId',
            },
            roles: [],
        };
        tree = new AuxCausalTree(storedTree(site(1)), {
            filter: (tree, atom) => {
                if (channel) {
                    return filterAtom(
                        <AuxCausalTree>tree,
                        atom,
                        () => channel.helper
                    );
                } else {
                    return true;
                }
            },
        });
        config = {
            config: {
                isBuilder: false,
                isPlayer: false,
                version: 'v1.0.0',
                versionHash: 'hash',
            },
            partitions: {
                shared: {
                    type: 'causal_tree',
                    id: 'auxId',
                    tree: tree,
                },
            },
        };
        await tree.root();

        channel = new AuxChannelImpl(user, device, config);
    });

    describe('init()', () => {
        it('should create a bot for the user', async () => {
            await channel.initAndWait();

            const userBot = channel.helper.userBot;
            expect(userBot).toBeTruthy();
            expect(userBot.tags).toMatchSnapshot();
        });

        it('should create a user dimension bot', async () => {
            uuidMock.mockReturnValue('dimensionBot');
            await channel.initAndWait();

            const dimensionBot = channel.helper.botsState['dimensionBot'];
            expect(dimensionBot).toBeTruthy();
            expect(dimensionBot.tags).toMatchSnapshot();
        });

        it('should create the globals bot', async () => {
            await channel.initAndWait();

            const globals = channel.helper.globalsBot;
            expect(globals).toBeTruthy();
            expect(globals.tags).toMatchSnapshot();
        });

        it('should load the builder aux file', async () => {
            channel = new AuxChannelImpl(
                user,
                device,
                merge({}, config, {
                    config: {
                        builder: JSON.stringify({
                            builder: createBot('builder', {
                                abc: 'def',
                                builderVersion: 0,
                            }),
                        }),
                    },
                })
            );
            await channel.initAndWait();

            const builderBot = channel.helper.botsState['builder'];
            expect(builderBot).toMatchObject({
                id: 'builder',
                tags: {
                    abc: 'def',
                    builderVersion: 0,
                },
            });
        });

        it('should not overwrite changes to builder from the aux file if the version is not newer', async () => {
            await tree.addBot(
                createBot('builder', {
                    different: true,
                    builderVersion: 2,
                })
            );

            channel = new AuxChannelImpl(
                user,
                device,
                merge({}, config, {
                    config: {
                        builder: JSON.stringify({
                            builder: createBot('builder', {
                                abc: 'def',
                                builderVersion: 2,
                            }),
                        }),
                    },
                })
            );
            await channel.initAndWait();

            const builderBot = channel.helper.botsState['builder'];
            expect(builderBot).toMatchObject({
                id: 'builder',
                tags: {
                    different: true,
                    builderVersion: 2,
                },
            });
        });

        it('should overwrite changes to builder from the aux file if the version is newer', async () => {
            await tree.addBot(
                createBot('builder', {
                    different: true,
                    builderVersion: 2,
                })
            );

            channel = new AuxChannelImpl(
                user,
                device,
                merge({}, config, {
                    config: {
                        builder: JSON.stringify({
                            builder: createBot('builder', {
                                abc: 'def',
                                builderVersion: 3,
                            }),
                        }),
                    },
                })
            );
            await channel.initAndWait();

            const builderBot = channel.helper.botsState['builder'];
            expect(builderBot).toMatchObject({
                id: 'builder',
                tags: {
                    different: true,
                    abc: 'def',
                    builderVersion: 3,
                },
            });
        });

        it('should allow users with the admin role', async () => {
            config.config.isBuilder = true;
            await tree.addBot(
                createBot(GLOBALS_BOT_ID, {
                    'aux.designers': ['notusername'],
                })
            );

            let messages: AuthorizationMessage[] = [];
            channel.onConnectionStateChanged.subscribe(m => {
                if (m.type === 'authorization') {
                    messages.push(m);
                }
            });

            device.roles.push(ADMIN_ROLE);
            await channel.init();

            for (let i = 0; i < 100; i++) {
                await Promise.resolve();
            }

            expect(messages).toEqual([
                {
                    type: 'authorization',
                    authorized: true,
                },
            ]);
        });

        it('should allow users with the server role', async () => {
            config.config.isBuilder = true;
            await tree.addBot(
                createBot(GLOBALS_BOT_ID, {
                    'aux.designers': ['notusername'],
                })
            );

            let messages: AuthorizationMessage[] = [];
            channel.onConnectionStateChanged.subscribe(m => {
                if (m.type === 'authorization') {
                    messages.push(m);
                }
            });

            device.roles.push(SERVER_ROLE);
            await channel.init();

            for (let i = 0; i < 100; i++) {
                await Promise.resolve();
            }

            expect(messages).toEqual([
                {
                    type: 'authorization',
                    authorized: true,
                },
            ]);
        });

        it('should not error if the tree does not have a root atom', async () => {
            tree = new AuxCausalTree(storedTree(site(1)));
            config = {
                config: {
                    isBuilder: false,
                    isPlayer: false,
                    version: 'v1.0.0',
                    versionHash: 'hash',
                },
                partitions: {
                    shared: {
                        type: 'causal_tree',
                        id: 'auxId',
                        tree: tree,
                    },
                },
            };
            channel = new AuxChannelImpl(user, device, config);

            await channel.initAndWait();
        });

        it('should error if unable to construct a partition', async () => {
            tree = new AuxCausalTree(storedTree(site(1)));
            config = {
                config: {
                    isBuilder: false,
                    isPlayer: false,
                    version: 'v1.0.0',
                    versionHash: 'hash',
                },
                partitions: {
                    shared: {
                        type: 'remote_causal_tree',
                        id: 'auxId',
                        host: 'host',
                        treeName: 'treeName',
                    },
                },
            };
            channel = new AuxChannelImpl(user, device, config);

            await expect(channel.initAndWait()).rejects.toEqual(
                new Error('[BaseAuxChannel] Unable to build partition: shared')
            );
        });

        it('should keep dimensions in users that define a dimension', async () => {
            await tree.addBot(
                createBot('user1', {
                    auxPlayerName: 'user',
                    auxDimensionConfig: `_user_user_1`,
                })
            );

            await channel.initAndWait();

            const userBot = channel.helper.botsState['user1'];
            expect(userBot).toBeTruthy();
            expect(userBot.tags).toEqual({
                auxPlayerName: 'user',
                auxDimensionConfig: '_user_user_1',
            });
        });
    });

    describe('onUniverseAction()', () => {
        it('should send new bot atoms through the onUniverseAction() filter', async () => {
            await channel.initAndWait();
            await tree.updateBot(channel.helper.globalsBot, {
                tags: {
                    onUniverseAction: `@
                        if (that.action.type === 'add_bot') {
                            action.reject(that.action);
                        }
                    `,
                },
            });

            const a = atom(atomId(2, 100), tree.weave.atoms[0].id, bot('test'));
            const { rejected } = await tree.add(a);

            expect(rejected).toEqual({
                atom: a,
                reason: 'rejected_by_filter',
            });
        });

        it('should send delete bot atoms through the onUniverseAction() filter', async () => {
            await channel.initAndWait();
            await tree.updateBot(channel.helper.globalsBot, {
                tags: {
                    onUniverseAction: `@
                        if (that.action.type === 'remove_bot') {
                            action.reject(that.action);
                        }
                    `,
                },
            });

            const { added } = await tree.addBot(createBot('test'));

            const a = atom(atomId(2, 100), added[0].id, del());
            const { rejected } = await tree.add(a);

            expect(rejected).toEqual({
                atom: a,
                reason: 'rejected_by_filter',
            });
        });

        it('should send update tag atoms through the onUniverseAction() filter', async () => {
            await channel.initAndWait();
            await tree.updateBot(channel.helper.globalsBot, {
                tags: {
                    onUniverseAction: `@
                        if (that.action.type === 'update_bot') {
                            action.reject(that.action);
                        }
                    `,
                },
            });

            const { added } = await tree.addBot(createBot('test'));

            const a1 = atom(atomId(2, 100), added[0].id, tag('abc'));
            const { rejected: rejected1 } = await tree.add(a1);

            const a2 = atom(atomId(2, 101), a1.id, value(123));
            const { rejected: rejected2 } = await tree.add(a2);

            expect(rejected1).toBe(null);
            expect(rejected2).toEqual({
                atom: a2,
                reason: 'rejected_by_filter',
            });
        });

        it('should send delete tag atoms through the onUniverseAction() filter', async () => {
            await channel.initAndWait();
            await tree.updateBot(channel.helper.globalsBot, {
                tags: {
                    onUniverseAction: `@
                        if (that.action.type === 'update_bot') {
                            action.reject(that.action);
                        }
                    `,
                },
            });

            const { added } = await tree.addBot(createBot('test'));

            const a1 = atom(atomId(2, 100), added[0].id, tag('abc'));
            const { rejected: rejected1 } = await tree.add(a1);

            const a2 = atom(atomId(2, 101), a1.id, del());
            const { rejected: rejected2 } = await tree.add(a2);

            expect(rejected1).toBe(null);
            expect(rejected2).toEqual({
                atom: a2,
                reason: 'rejected_by_filter',
            });
        });
    });

    describe('sendEvents()', () => {
        it('should send remote events to _sendRemoteEvents()', async () => {
            await channel.initAndWait();

            await channel.sendEvents([
                {
                    type: 'remote',
                    event: botAdded(createBot('def')),
                },
                botAdded(createBot('test')),
                {
                    type: 'remote',
                    event: botAdded(createBot('abc')),
                },
            ]);

            expect(channel.remoteEvents).toEqual([
                remote(botAdded(createBot('def'))),
                remote(botAdded(createBot('abc'))),
            ]);
        });

        it('should send device events to onDeviceEvents', async () => {
            await channel.initAndWait();

            let deviceEvents: DeviceAction[] = [];
            channel.onDeviceEvents.subscribe(e => deviceEvents.push(...e));

            await channel.sendEvents([
                {
                    type: 'device',
                    device: {
                        claims: {
                            [USERNAME_CLAIM]: 'username',
                            [DEVICE_ID_CLAIM]: 'deviceId',
                            [SESSION_ID_CLAIM]: 'sessionId',
                        },
                        roles: ['role'],
                    },
                    event: botAdded(createBot('def')),
                },
                botAdded(createBot('test')),
                {
                    type: 'device',
                    device: null,
                    event: botAdded(createBot('abc')),
                },
            ]);

            expect(deviceEvents).toEqual([
                {
                    type: 'device',
                    device: {
                        claims: {
                            [USERNAME_CLAIM]: 'username',
                            [DEVICE_ID_CLAIM]: 'deviceId',
                            [SESSION_ID_CLAIM]: 'sessionId',
                        },
                        roles: ['role'],
                    },
                    event: botAdded(createBot('def')),
                },
                {
                    type: 'device',
                    device: null,
                    event: botAdded(createBot('abc')),
                },
            ]);
        });

        describe('load_space', () => {
            it('should handle load_space events', async () => {
                await channel.initAndWait();

                await channel.sendEvents([
                    {
                        type: 'load_space',
                        space: 'tempLocal',
                        config: <MemoryPartitionConfig>{
                            type: 'memory',
                            initialState: {
                                abc: createBot('abc'),
                            },
                        },
                    },
                ]);

                const { abc } = channel.helper.botsState;
                expect(abc).toEqual(createBot('abc', {}, 'tempLocal'));
            });

            it('should not overwrite existing spaces', async () => {
                await channel.initAndWait();

                await channel.sendEvents([
                    {
                        type: 'load_space',
                        space: 'shared',
                        config: <MemoryPartitionConfig>{
                            type: 'memory',
                            initialState: {
                                abc: createBot('abc'),
                            },
                        },
                    },
                ]);

                const { abc } = channel.helper.botsState;
                expect(abc).toBeUndefined();
            });
        });
    });

    describe('formulaBatch()', () => {
        it('should send remote events', async () => {
            await channel.initAndWait();

            await channel.formulaBatch(['server.browseHistory()']);

            expect(channel.remoteEvents).toEqual([remote(browseHistory())]);
        });
    });

    describe('search', () => {
        it('should convert errors to copiable values', async () => {
            await channel.initAndWait();

            const result = await channel.search('throw new Error("abc")');

            expect(result).toEqual({
                success: false,
                extras: expect.any(Object),
                error: 'Error: abc',
                logs: expect.any(Array),
            });
        });
    });

    describe('export()', () => {
        beforeEach(async () => {
            config = {
                config: {
                    isBuilder: false,
                    isPlayer: false,
                    version: 'v1.0.0',
                    versionHash: 'hash',
                },
                partitions: {
                    shared: {
                        type: 'causal_tree',
                        id: 'auxId',
                        tree: tree,
                    },
                    tempLocal: {
                        type: 'memory',
                        initialState: {
                            def: createBot('def'),
                        },
                    },
                    private: {
                        type: 'memory',
                        initialState: {
                            private: createBot('private'),
                        },
                        private: true,
                    },
                },
            };
            await tree.root();

            channel = new AuxChannelImpl(user, device, config);
        });

        it('should only export public bots', async () => {
            uuidMock.mockReturnValue('dimensionBot');
            await channel.initAndWait();

            await channel.sendEvents([botAdded(createBot('test'))]);

            const exported = await channel.export();

            expect(exported).toEqual({
                version: 1,
                state: {
                    config: expect.any(Object),
                    dimensionBot: expect.any(Object),
                    userId: expect.any(Object),
                    test: createBot('test', {}, 'shared'),
                    def: createBot('def', {}, 'tempLocal'),
                },
            });
        });

        it('should inlcude the ID, tags, and space properties', async () => {
            uuidMock.mockReturnValue('dimensionBot');
            await channel.initAndWait();

            await channel.sendEvents([botAdded(createBot('test'))]);

            const exported = await channel.export();

            expect(exported).toEqual({
                version: 1,
                state: {
                    config: expect.any(Object),
                    dimensionBot: expect.any(Object),
                    userId: expect.any(Object),
                    test: createBot('test', {}, 'shared'),
                    def: createBot('def', {}, 'tempLocal'),
                },
            });
        });
    });
    // describe('forkAux()', () => {
    //     it('should call fork on the partitions', async () => {
    //         await channel.initAndWait();

    //         await channel.forkAux('test2');

    //     });
    // });
});

class AuxChannelImpl extends BaseAuxChannel {
    remoteEvents: RemoteAction[];

    private _device: DeviceInfo;
    constructor(user: AuxUser, device: DeviceInfo, config: AuxConfig) {
        super(user, config, {});
        this._device = device;
        this.remoteEvents = [];
    }

    protected async _sendRemoteEvents(events: RemoteAction[]): Promise<void> {
        this.remoteEvents.push(...events);
    }

    protected _createPartition(config: PartitionConfig): Promise<AuxPartition> {
        return createAuxPartition(
            config,
            createLocalCausalTreePartitionFactory({}, this.user, this._device),
            cfg => createMemoryPartition(cfg)
        );
    }
}
