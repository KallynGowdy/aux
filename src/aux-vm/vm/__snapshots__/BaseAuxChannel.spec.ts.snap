// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseAuxChannel init() should create a bot for the user 1`] = `
Object {
  "_auxUser": "username",
  "_auxUserChannelsContext": "_user_username_simulations",
  "_auxUserInventoryContext": "_user_username_inventory",
  "_auxUserMenuContext": "_user_username_menu",
  "aux-users": true,
}
`;

exports[`BaseAuxChannel init() should create a user context bot 1`] = `
Object {
  "aux.context": "aux-users",
  "auxContextVisualize": true,
}
`;

exports[`BaseAuxChannel init() should create the globals bot 1`] = `
Object {
  "auxDestroyable": false,
  "auxVersion": 1,
}
`;
